File: 0468_validate_IP/is_valid_IP.py
import re


def valid_IP_address(IP: str) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    if is_valid_IP4(IP):
        return "IPv4"
    if is_valid_IP6(IP):
        return "IPv6"
    return "Neither"


def is_valid_IP4(IP: str) -> bool:
    if IP.count('.') != 3:
        return False
    split = IP.split('.')
    if any(not x or
           not x.isdigit() or
           int(x) > 255 or
           (x[0] == "0" and len(x) > 1) for x in split):
        return False
    return True


def is_valid_IP6(IP: str) -> bool:
    if IP.count(':') != 7:
        return False
    split = IP.split(':')
    hexa = "0123456789ABCDEFabcdef"
    if any(not x or
           len(x) > 4 or
           any(c not in hexa for c in x) for x in split):
        return False
    return True


def is_valid_regex(IP: str) -> str:
    if re.search(r'^((25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)\.){3}(25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)$', IP):
        return 'IPv4'
    elif re.search(r'^([\da-f]{1,4}:){7}[\da-f]{1,4}$', IP, re.I):
        return 'IPv6'
    return 'Neither'

File: 0468_validate_IP/is_valid_IP.java
class Solution {
    public String validIPAddress(String IP) {
        if (isValidIPv4(IP)) {
            return "IPv4";
        }
        if (isValidIPv6(IP)) {
            return "IPv6";
        }
        return "Neither";
    }

     private boolean isValidIPv4(String IP) {
        return (isValidIPv4Length(IP) && isValidIPv4Format(IP));
    }

    private boolean isValidIPv4Length(String IP) {
        int cnt = 0;
        for (char ch : IP.toCharArray()) {
            if (ch == '.') {
                cnt++;
            }
        }
        return cnt == 3;
    }

    private boolean isValidIPv4Format(String IP) {
        String[] fields = IP.split("\\.");
        if (fields.length != 4) {
            return false;
        }
        for (String field : fields) {
            if (field.isEmpty() || field.length() > 3) {
                return false;
            }
            int sz = field.length();
            for (int i = 0; i < sz; ++i) {
                if (!Character.isDigit(field.charAt(i))) {
                    return false;
                }
            }
            int num = Integer.valueOf(field);
            if (!String.valueOf(num).equals(field) || num < 0 || num > 255) {
                return false;
            }
        }
        return true;
    }
        
    private boolean isValidIPv6(String IP) {
        return isValidIPv6Length(IP) && isValidIPv6Format(IP);
    }

    private boolean isValidIPv6Length(String IP) {
        int cnt = 0;
        for (char ch : IP.toCharArray()) {
            if (ch == ':') {
                cnt++;
            }
        }
        return cnt == 7;
    }

    private boolean isValidIPv6Format(String IP) {
        String[] fields = IP.split(":");
        if (fields.length != 8) {
            return false;
        }
        for (String field : fields) {
            if (field.isEmpty() || field.length() > 4) {
                return false;
            }
            int sz = field.length();
            for (int i = 0; i < sz; ++i) {
                if (!isValidIPv6Char(field.charAt(i))) {
                    return false;
                }       
            }
        }
        return true;
    }

    private boolean isValidIPv6Char(char ch) {
        return (ch >= '0' && ch <= '9') || 
               (ch >= 'A' && ch <= 'F') || 
               (ch >= 'a' && ch <= 'f');
    }
}